client/src/client.rs

```diff
@@ -939,7 +946,27 @@ impl<B, E, Block, RA> Client<B, E, Block, RA> where
 		Ok(())
 	}
 
-	fn notify_imported(&self, notify_import: ImportSummary<Block>) -> sp_blockchain::Result<()> {
+	fn notify_imported(
+		&self,
+		notify_import: Option<ImportSummary<Block>>,
+	) -> sp_blockchain::Result<()> {
+		let notify_import = match notify_import {
+			Some(notify_import) => notify_import,
+			None => {
+				// cleanup any closed import notification sinks since we won't
+				// be sending any notifications below which would remove any
+				// closed sinks. this is necessary since during initial sync we
+				// won't send any import notifications which could lead to a
+				// temporary leak of closed/discarded notification sinks (e.g.
+				// from consensus code).
+				self.import_notification_sinks
+					.lock()
+					.retain(|sink| !sink.is_closed());
```

Comment:
It feels a bit wasteful to iterate over all sinks on EVERY block import now, but I suppose the list should have at most tens of elements, right?

Reply:
Yes

Reply:
At first I was thinking of doing this only every `N` blocks, but since compared to block import time this operation should be one or two orders of magnitude faster I think it is OK. Yeah, the number of subscribers should be in the tens.